File: findLadders.cpp
/**
 * @author f4prime
 * @email zzz136454872@163.com
 * @aim a plain model for leetcode
 */

#include<iostream>
#include<string>
#include<vector>
#include<algorithm>
#include<map>
#include<stack>
#include<queue>

using namespace std;

//显示中间结果
void print(vector<int> a)
{
    for(int i=0;i<(int)a.size();i++)
        cout<<a[i]<<" ";
    cout<<endl;
}

void print(vector<vector<int>> a)
{
    for(int i=0;i<(int)a.size();i++)
    {
        for(int j=0;j<(int)a[i].size();j++)
            cout<<a[i][j]<<" ";
        cout<<endl;
    }
}

void print(vector<bool> a)
{
    for(int i=0;i<(int)a.size();i++)
    {
        cout<<i<<" ";
        if(a[i])
            cout<<"true ";
        else
            cout<<"false ";
    }
    cout<<endl;
}

void print(bool a[],int len)
{
    for(int i=0;i<len;i++)
    {
        cout<<i<<" ";
        if(a[i])
            cout<<"true ";
        else
            cout<<"false ";
    }
    cout<<endl;
}

//#define testMod

#ifdef testMod
void test()
{
    
}
#endif

#ifndef testMod
class Solution {
    public:
        vector<vector<string>> findLadders(string beginWord, string endWord, vector<string>& wordList) {
            start=beginWord;
            stop=endWord;
            string now=start;
            vector<bool> log(wordList.size(),false);
            this->list=wordList;
            return findpath(now);
        }

        vector<vector<string>> findpath(string now,vector<bool> log) {
            cout<<now<<endl;
            vector<string> tmp;
            vector<vector<string>> out;
            string t=now;
            if(now.compare(stop)==0)
            {
                tmp.push_back(stop);
                out.push_back(tmp);
                return out;
            }
            for(int i=0;i<(int)list.size();i++)
            {
                if(log[i]==false&&diff(now,list[i])==1)
                {
                    log[i]=true;
                    if(index!=list.end())
                    {
                        vector<vector<string>> l_out=findpath(log[i]);
                        for(auto vs:l_out)
                        {
                            vs.insert(vs.begin(),now);
                            out.insert(out.begin(),vs);
                        }
                    }
                }
            }
            return out;
        }

        int diff(string a,string b)
        {
            int out=0;
            for(int i=0;i<(int)a.size();i++)
            {
                if(a[i]!=b[i])
                    out++;
            }
            return out;
        }

    private:
        string start,stop;
        vector<string> list;
};
#endif

int main()
{
#ifdef testMod
    test();
#endif

#ifndef testMod
    Solution sl;
    vector<string> wordList={"hot","dot","dog","lot","log","cog"};
    string beginWord = "hit",endWord = "cog";
    vector<vector<string>> out=sl.findLadders(beginWord,endWord,wordList);
    for(auto vs:out)
    {
        for(string a:vs)
            cout<<a<<" ";
        cout<<endl;
    }
#endif 
    return 0;
}

File: findLadders.py
from collections import deque
from typing import List


# 不知道是哪个
class Solution1:
    def findLadders(self, beginWord: str, endWord: str,
                    wordList: List[str]) -> int:
        wordList.append(beginWord)

        if endWord not in wordList:
            return []
        dic = dict()
        levelLog = {word: 5000 for word in wordList}
        before = {word: [] for word in wordList}

        for word in wordList:
            for i in range(len(word)):
                key = word[:i] + '*' + word[i + 1:]

                if key in dic.keys():
                    dic[key].append(word)
                else:
                    dic[key] = [word]

        queue = [(beginWord, 1)]
        levelLog[beginWord] = 1
        stopLevel = 5000

        while len(queue) > 0:
            item = queue[0]
            del queue[0]
            level = item[1] + 1

            if level > stopLevel:
                break

            for i in range(len(item[0])):
                key = item[0][:i] + '*' + item[0][i + 1:]

                for neighbor in dic[key]:
                    if levelLog[neighbor] == 5000:
                        levelLog[neighbor] = level
                        before[neighbor].append(item[0])

                        if neighbor == endWord:
                            stopLevel = level
                        queue.append((neighbor, level))
                    elif levelLog[neighbor] == level and item[0] not in before[
                            neighbor]:
                        before[neighbor].append(item[0])

        def getAll(word):
            if word == beginWord:
                return [[beginWord]]
            out = []

            for b in before[word]:
                out += [path + [word] for path in getAll(b)]

            return out

        return getAll(endWord)


# beginWord = "hit"
# endWord = "cog"
# wordList = ["hot","dot","dog","lot","log","cog"]
#
# sl=Solution()
# print(sl.findLadders(beginWord,endWord,wordList))


class Solution:
    def findLadders(self, beginWord: str, endWord: str,
                    wordList: List[str]) -> List[str]:
        if endWord not in wordList:
            return []

        if beginWord not in wordList:
            wordList.append(beginWord)

        dic = dict()

        for word in wordList:
            for i in range(len(word)):
                key = word[:i] + '*' + word[i + 1:]

                if key in dic.keys():
                    dic[key].append(word)
                else:
                    dic[key] = [word]

        before = {k: "" for k in wordList}
        queue = deque([(beginWord, 0)])
        level = 0
        find = False

        while len(queue) > 0:
            word, level = queue.popleft()

            if level > 5000:
                return []

            for i in range(len(word)):
                key = word[:i] + '*' + word[i + 1:]

                for neighbor in dic[key]:
                    if neighbor == word:
                        continue

                    if before[neighbor] == "":
                        before[neighbor] = word
                    else:
                        continue

                    if neighbor == endWord:
                        find = True

                        break
                    queue.append((neighbor, level + 1))

                if find:
                    break

            if find:
                break

        if not find:
            return []
        res = []
        tmp = endWord

        while tmp != beginWord:
            res.append(tmp)
            tmp = before[tmp]
        res.append(tmp)

        return res[::-1]


beginWord = "hit"
endWord = "cog"
wordList = ["hot", "dot", "dog", "lot", "log", "cog"]
beginWord = "hit"
endWord = "cog"
wordList = ["hot", "dot", "dog", "lot", "log"]
beginWord = "hot"
endWord = "dog"
wordList = ["hot", "dog"]

print(Solution().findLadders(beginWord, endWord, wordList))
